src/config/loading/secret.rs

```diff
@@ -0,0 +1,341 @@
+struct ExecBackend {
+    pub command: Vec<String>,
+    #[serde(default = "default_timeout_secs")]
+    pub timeout: u64,
+}
+
+const fn default_timeout_secs() -> u64 {
+    5
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+struct ExecQuery {
+    version: String,
+    secrets: Vec<String>,
+}
+
+fn new_query(secrets: Vec<String>) -> ExecQuery {
+    ExecQuery {
+        version: "1.0".to_string(),
+        secrets,
+    }
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+struct ExecResponse {
+    value: Option<String>,
+    error: Option<String>,
+}
+
+#[typetag::serde(name = "exec")]
+impl SecretBackend for ExecBackend {
+    fn retrieve(
+        &mut self,
+        secret_keys: Vec<String>,
+        signal_rx: &mut signal::SignalRx,
+    ) -> crate::Result<HashMap<String, String>> {
+        let mut output = executor::block_on(tokio::time::timeout(
+            Duration::from_secs(self.timeout),
+            async {
+                loop {
+                    tokio::select! {
+                        biased;
+                        Ok(signal::SignalTo::Shutdown | signal::SignalTo::Quit) = signal_rx.recv() => {
+                            break Err("Secret retrieval was interrupted.".into());
+                        }
+                        output = query_backend(&self.command, new_query(secret_keys.clone())) => break output,
+                    }
+                }
+            },
+        ))??;
+        let mut secrets = HashMap::new();
+        for k in secret_keys.into_iter() {
+            if let Some(secret) = output.get_mut(&k) {
+                if let Some(e) = &secret.error {
+                    return Err(format!("secret for key '{}' was not retrieved: {}", k, e).into());
+                }
+                if let Some(v) = secret.value.take() {
+                    if v.is_empty() {
+                        return Err(format!("secret for key '{}' was empty", k).into());
+                    }
+                    secrets.insert(k, v);
+                } else {
+                    return Err(format!("secret for key '{}' was empty", k).into());
+                }
+            } else {
+                return Err(format!("secret for key '{}' was not retrieved", k).into());
+            }
+        }
+        Ok(secrets)
+    }
+}
+
+async fn query_backend(
+    cmd: &[String],
+    query: ExecQuery,
+) -> crate::Result<HashMap<String, ExecResponse>> {
+    let command = &cmd[0];
+    let mut command = Command::new(command);
+
+    if cmd.len() > 1 {
+        command.args(&cmd[1..]);
+    };
+
+    command.kill_on_drop(true);
+    command.stderr(std::process::Stdio::piped());
+    command.stdin(std::process::Stdio::piped());
+    command.stdout(std::process::Stdio::piped());
+
+    let mut child = command.spawn()?;
+    let mut stdin = child.stdin.take().ok_or("unable to acquire stdin")?;
+
+    let query = serde_json::to_vec(&query)?;
+
+    tokio::spawn(async move { stdin.write_all(&query).await });
```

Comment:
I see you are handling the timeout above, but that won't actually kill this spawned task. I think maybe you should pass the timeout into this function and use it here to abort the task at timeout and return an error.